File: competitive-programming/src/edu/algos/li01_array/medium/L238_ProductOfArrayExceptSelf.java
package edu.algos.li01_array.medium;

import java.util.Arrays;

public class L238_ProductOfArrayExceptSelf {

	public static void main(String[] args) {
		L238_ProductOfArrayExceptSelf obj = new L238_ProductOfArrayExceptSelf();
		System.out.println(Arrays.toString(obj.productExceptSelf(new int[] {1,2,3,4})));
	}

    // Problem: https://leetcode.com/problems/product-of-array-except-self/
    // Idea: Track prefix and suffix products. Result is prefix prod*suffix prod.
	public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        int prefixProd = 1;
        int postfixProd = 1;
        for(int i=0; i<nums.length; i++){
            result[i] = prefixProd;    
            prefixProd *= nums[i];
        }
        
        for(int i=nums.length-1; i>=0; i--){
            result[i] *= postfixProd;
            postfixProd *= nums[i];
        }
        
        return result;
    }
}

File: dsa/topics/01_arrays/python/l238_product_of_array_except_self.py
# L238 - Product of Array Except Self
#
# Problem: https://leetcode.com/problems/product-of-array-except-self/
# Idea: Track prefix and suffix products. Result is prefix_prod * suffix_prod.
# Time: O(n)
# Space: O(1) - excluding the output array


def product_except_self(nums):
    n = len(nums)
    res = [1] * n

    pref_prod = 1
    for i in range(n):
        res[i] = pref_prod
        pref_prod *= nums[i]

    suf_prod = 1
    for i in range(n-1, -1, -1):
        res[i] *= suf_prod
        suf_prod *= nums[i]

    return res


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    print(product_except_self(nums))  # [24, 12, 8, 6]
